### ISSUE TITLE
✨ Add reverse-mode automatic differentiation

### ISSUE THREAD

**Issue by rjkilpatrick** [rjkilpatrick/NumD#7]
**Is your feature request related to a problem? Please describe.**
Many problems in the sciences use gradient descent to solve a problem, e.g. [NMF](https://en.wikipedia.org/wiki/Non-negative_matrix_factorization), [Deep Learning](https://en.wikipedia.org/wiki/Deep_learning).
Manually calculating this gradient is tedious, so we aim to automatically determine the gradients from a computation graph.

**Describe the solution you'd like**
Reverse-mode (and possibly forward mode) automatic differentiation.

To find the weights and biases that mean our forward model is consistent with our measurements of y,
![yHat = Wx + b](https://latex.codecogs.com/svg.latex?%5Cmathbf%7B%5Chat%7By%7D%7D%20%3D%20%5Cmathbf%7BW%7D%20%5Cmathbf%7Bx%7D%20&plus;%20%5Cmathbf%7Bb%7D)
by minimizing a loss function of
![loss = l2 norm of (y - yHat)](https://latex.codecogs.com/svg.latex?%5Cmathcal%7BL%7D%20%3D%20%5C%7C%20%5Cmathbf%7B%5Chat%7By%7D%7D%20-%20%5Cmathbf%7By%7D%20%5C%7C%5E2_2)

```d
import numd;
import numd.allocation;
import numd.linalg : matrixMultiply, norm;

auto x = ones(5); // Input tensor
auto y = zeros(3);  // Ground truth

auto w = rand(5, 3); // Denote some kind of gradient required since we want it to update in value
auto b = rand(3); // Denote some kind of gradient required since we want it to update in value
auto yHat = matrixMultiply(x, w) + b;

// This is the scalar we want to minimize
auto loss = norm(yHat - y);

loss.backward(); // Calculate all gradients from before this point in the graph
```
`w.grad` is a property populated with a gradient ![partial d L / partial d W](https://latex.codecogs.com/svg.latex?%5Cinline%20%5Cfrac%7B%5Cpartial%20%5Cmathcal%7BL%7D%7D%7B%5Cpartial%20%5Cmathbf%7BW%7D%7D), and `b.grad` is populated with the gradient ![partial d L / partial d b](https://latex.codecogs.com/svg.latex?%5Cinline%20%5Cfrac%7B%5Cpartial%20%5Cmathcal%7BL%7D%7D%7B%5Cpartial%20%5Cmathbf%7Bb%7D%7D)

We can then update the weights and biases by
```d
auto learningRate = 1.e-3;
w -= learningRate * w.grad;
b -= learningRate * b.grad;
```

**Describe alternatives you've considered**
Using PyTorch C++ API with a Dlang wrapper, but this is cumbersome.

**Additional context**
Mirror the [PyTorch API](https://pytorch.org/tutorials/beginner/basics/autogradqs_tutorial.html)
